<?xml version="1.0" encoding="utf-8"?>
<launch>

  <arg name="type" default="quad" />
  <arg name="transformation" default="true" />
  <arg name="se2" default="false" />
  <arg name="horizontal_gap" default="false" />

  <arg name="load_path_flag" default="true" />
  <arg name="save_path_flag" default="true" />
  <arg name="play_path_flag" default="false" />
  <arg name="realtime_path_flag" default="false" />
  <arg name="path_file_name" default="dragon_planning_log.txt" />

  ###########  Robot Model  ###########
  <group if="$(arg play_path_flag)">
    <include file="$(find dragon)/launch/robot_model.launch" >
      <arg name="type" value="$(arg type)" />
      <arg name="headless" value="true" />
      <arg name="need_joint_state" value="false"/>
    </include >

    <node name="rviz" pkg="rviz" type="rviz"
          args="-d $(find gap_passing)/config/data_replay.rviz" respawn="true" if="$(arg realtime_path_flag)"/>

    <node name="rviz" pkg="rviz" type="rviz"
          args="-d $(find gap_passing)/config/planning.rviz" respawn="true" unless="$(arg realtime_path_flag)"/>
  </group>

  ########## Planning for gap passing ###########
  <param name="/robot_description_semantic" textfile="$(find gap_passing)/config/dragon/$(arg type)/moveit.srdf" />

  <node name="motion_planning_node" pkg="gap_passing" type="se3_motion_planning_node" respawn="false" output="screen">
    <param name="play_path_flag" type="bool" value="$(arg play_path_flag)"/>
    <param name="save_path_flag" type="bool" value="$(arg save_path_flag)" />
    <param name="load_path_flag" type="bool" value="$(arg load_path_flag)" />
    <param name="realtime_path_flag" type="bool" value="$(arg realtime_path_flag)" />

    <param name="file_name" value="$(find gap_passing)/path_log/dragon/$(arg path_file_name)" />

    <param name="planning_mode" value="1" unless="$(arg transformation)"/>
    <param name="planning_mode" value="2" if="$(arg transformation)"/>
    <param name="locomotion_mode" value="0" if="$(arg se2)"/>
    <param name="locomotion_mode" value="1" unless="$(arg se2)"/> <!-- se3 -->
    <param name="gap_type" value="0" if="$(arg horizontal_gap)"/>
    <param name="gap_type" value="1" unless="$(arg horizontal_gap)"/>

    <param name="solving_time_limit" type="double" value="18000"/>
    <param name="motion_sequence_rate" value="10"/>

    <param name="coefficient_rate" value="0.05"/>
    <param name="ompl_mode" value="0"/> <!-- RRT START:0 -->
    <param name="state_validity_check_res" value="0.005"/>
    <param name="valid_segment_count_factor" value="20"/> <!--20-->

    <param name="file_state_offset_x" type="double" value="0.0" />
    <param name="file_state_offset_y" type="double" value="0.0" />
    <param name="file_state_offset_z" type="double" value="0.0" />

    <!-- for specific param -->
    <rosparam file="$(find dragon)/config/$(arg type)/AttitudeControl.yaml" command="load" />
    <rosparam file="$(find gap_passing)/config/dragon/$(arg type)/GapWithTransformation.yaml" command="load" if="$(arg transformation)"/>
    <rosparam file="$(find gap_passing)/config/dragon/$(arg type)/GapWithoutTransformation.yaml" command="load" unless="$(arg transformation)"/>
    <rosparam file="$(find gap_passing)/config/gap/HorizontalGapNoOverlap.yaml" command="load" if="$(arg horizontal_gap)"/>
    <rosparam file="$(find gap_passing)/config/gap/VerticalGap.yaml" command="load" unless="$(arg horizontal_gap)"/>
  </node>
</launch>
